### PR TITLE
Feature/show DRM violations in dialog

### PR DESCRIPTION
**Pull request by PenghaiZhang**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [x] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This PR adds the support for listing DRM violations for unauthorised Items and attachments. The main change is to update the logic of DRM check in `DrmHelper`. This change results in function `createDrmDialog` becoming an async function, and therefore the `useEffect` which uses this function gets slightly changed as well. 

I tried to implement a custom hook `useAsyncEffect` but stopped after a couple hours experiments as I thought the next DRM ticket is more important.

https://user-images.githubusercontent.com/47203811/127578666-fe879e2a-1ea5-4687-9571-f3371a170598.mp4

### CODE REVIEW

File: react-front-end/tsrc/drm/DrmHelper.tsx [hunk 1]
```diff
@@ -66,8 +71,40 @@ export const createDrmDialog = (
...
+    );
   } else {
     drmProtectedHandler();
     return;
   }
 };
+
+// Call the API to list DRM violations in a TaskEither where Left is why fail to
+// list violations and Right is the retrieved violation.
+const listDrmViolationsInTask = async (
```
Comment (PenghaiZhang) [inline react-front-end/tsrc/drm/DrmHelper.tsx:92]: This function was in the module but I moved it to the helper.

File: react-front-end/__tests__/tsrc/search/components/GallerySearchResult.test.tsx [hunk 2]
```diff
@@ -120,5 +134,26 @@ describe("<GallerySearchResult />", () => {
         });
       }
     );
+
+    it.each([
+      ["item", languageStrings.searchpage.gallerySearchResult.viewItem],
+      ["attachment", languageStrings.searchpage.gallerySearchResult.ariaLabel],
```
Comment (edalex-ian) [inline react-front-end/__tests__/tsrc/search/components/GallerySearchResult.test.tsx:140]: The `languageStrings` have been de-structured up the top, so:

```suggestion
      ["item", viewItem],
      ["attachment", ariaLabel],
```

That `ariaLabel` though keeps grabbing my attention. At somepoint we may want to refactor that with a more useful name - or maybe just in the de-structured context it loses meaning.

File: react-front-end/__tests__/tsrc/search/components/GallerySearchResult.test.tsx [hunk 3]
```diff
@@ -120,5 +134,26 @@ describe("<GallerySearchResult />", () => {
...
+    ])(
+      "shows a dialog to list DRM violations for %s",
+      async (_: string, galleryEntryLabel: string) => {
+        const { getByLabelText } = await renderGallery([
+          galleryDrmUnauthorisedItem,
+        ]);
+        await act(async () => {
+          await userEvent.click(getByLabelText(galleryEntryLabel));
+        });
```
Comment (edalex-ian) [inline react-front-end/__tests__/tsrc/search/components/GallerySearchResult.test.tsx:149]: There's a lot of boilerplate here. I wonder if we need a general (available for all tests) function like:

```typescript
const asyncClick = (element: Element, init?: MouseEventInit, {skipHover, clickCount}?: clickOptions): void =>
  act(async () => {
    await userEvent.click(element);
  });
```

And then we'd just be able to have:

```suggestion
        await asyncClick(getByLabelText(galleryEntryLabel));
```

I wonder too if `asyncClick` is the best name. Maybe `clickThenWait` as really this is all about clicking a button and then waiting for React to do it's state updates etc. Hmmm... :thinking:

File: react-front-end/__tests__/tsrc/search/components/SearchResult.test.tsx [hunk 4]
```diff
@@ -537,5 +540,23 @@ describe("<SearchResult/>", () => {
         });
       }
     );
+
+    it.each([
+      ["item", "DRM Item"],
+      ["attachment", "DRM Attachment"],
```
Comment (edalex-ian) [inline react-front-end/__tests__/tsrc/search/components/SearchResult.test.tsx:546]: The string literals of `"DRM Item"` and `"DRM Attachment"` are now in three places and ideally should all be linking to the same thing. Perhaps over in the mock we need to set a constant, and then use that to tie them altogether.

File: react-front-end/tsrc/drm/DrmHelper.tsx [hunk 5]
```diff
@@ -66,8 +71,40 @@ export const createDrmDialog = (
...
+    TE.map<OEQ.Drm.DrmViolation, string>(({ violation }) => violation)
+  )();
+
+  return pipe(
+    violation,
+    E.getOrElse(
+      (error) => `Failed to list DRM violations due to ${error.message}`
+    )
+  );
+};
```
Comment (edalex-ian) [inline react-front-end/tsrc/drm/DrmHelper.tsx:110]: Does this need to be broken up, or can it just be:

```suggestion
): Promise<string> =>
  await pipe(
    TE.tryCatch<Error, OEQ.Drm.DrmViolation>(
      () => listDrmViolations(uuid, version),
      (error) => new Error(`${error}`)
    ),
    TE.map<OEQ.Drm.DrmViolation, string>(({ violation }) => violation)
    TE.getOrElse(
      (error) => `Failed to list DRM violations due to ${error.message}`
    )
  )();
```

Are are the issues with the async/await?

File: react-front-end/tsrc/drm/DrmHelper.tsx [hunk 6]
```diff
@@ -66,8 +71,40 @@ export const createDrmDialog = (
         open={!!drmProtectedHandler}
       />
     );
+  } else if (!isAuthorised) {
```
Comment (edalex-ian) [inline react-front-end/tsrc/drm/DrmHelper.tsx:74]: I wonder if we should more so go with a 'deny all' default for the sake of robustness. To demonstrate with pseudo code:

```
if isAuthorised
  if termsAccepted
    call handler
  else
    show DrmAcceptanceDialog
else
  show Violations dialog
```

**Review by edalex-ian — CHANGES_REQUESTED** [review]
Nice @PenghaiZhang , this has come together rather cleanly.

Just a couple of minor requests, and we probably need to check the wording on the dialog with @cathfitz

File: react-front-end/__tests__/tsrc/search/components/GallerySearchResult.test.tsx (hunk 2 above)
Comment (PenghaiZhang) [reply react-front-end/__tests__/tsrc/search/components/GallerySearchResult.test.tsx:140]: Perhaps rename to `viewAttachment` ?

File: react-front-end/__tests__/tsrc/search/components/GallerySearchResult.test.tsx (hunk 3 above)
Comment (PenghaiZhang) [reply react-front-end/__tests__/tsrc/search/components/GallerySearchResult.test.tsx:149]: As discussed, we leave this for now and try to do it maybe in next tech debt.

**Pull request merged**